File: src/app/ics.py
from datetime import datetime, timedelta, timezone
from typing import Optional, Dict, Tuple
from icalendar import Calendar, Event, Alarm
from dateutil import parser as dtparser
try:
    from .constants import (
        EMOJI_STATUS,
        STATUS_CANONICAL_VARIANTS,
        STATUS_EMOJI,
    )
except ImportError:
    from constants import (  # type: ignore
        EMOJI_STATUS,
        STATUS_CANONICAL_VARIANTS,
        STATUS_EMOJI,
    )


DEFAULT_TIMED_EVENT_DURATION = timedelta(minutes=0)


def _compose_description(
    *,
    category: Optional[str],
    description: Optional[str],
) -> str:
    if description:
        return description
    if category:
        return f"Category: {category}"
    return ""


def _extract_summary_status(summary: str) -> Tuple[Optional[str], str]:
    if not summary:
        return None, ""
    head, sep, tail = summary.partition(' ')
    if sep and head in EMOJI_STATUS:
        return EMOJI_STATUS[head], tail.lstrip()
    first_char = summary[0]
    if first_char in EMOJI_STATUS:
        return EMOJI_STATUS[first_char], summary[1:].lstrip()
    return None, summary


_STATUS_PREFIXES_LOWER = sorted(
    {
        variant.strip().lower()
        for variants in STATUS_CANONICAL_VARIANTS.values()
        for variant in variants
        if variant
    },
    key=len,
    reverse=True,
)


def _clean_summary_title(title: Optional[str]) -> str:
    if not title:
        return ""
    working = title.lstrip()
    if working and working[0] in EMOJI_STATUS:
        working = working[1:].lstrip()
    lowered = working.lower()
    for prefix in _STATUS_PREFIXES_LOWER:
        if lowered.startswith(prefix):
            remainder = working[len(prefix):]
            while remainder and remainder[0] in (" ", "-", "–", "—", ":", "|"):
                remainder = remainder[1:]
            return remainder or working[len(prefix):].strip()
    return working


def _parse_description_fields(text: str) -> tuple[dict[str, str], Optional[str]]:
    headers = {}
    body = None
    header_text = text
    if "\n\n" in text:
        header_text, body_text = text.split("\n\n", 1)
        body = body_text.strip() or None
    header_candidates = []
    if "\n" in header_text:
        header_candidates.extend(
            line.strip() for line in header_text.splitlines() if line.strip()
        )
    else:
        header_candidates.extend(
            part.strip() for part in header_text.split("|") if part.strip()
        )
    for item in header_candidates:
        if ":" not in item:
            continue
        key, value = item.split(":", 1)
        headers[key.strip()] = value.strip()
    if body is None and "Description:" in headers:
        body = headers.get("Description") or None
    return headers, body


def build_uid(notion_id: str) -> str:
    return f"notion-{notion_id}@sync"


def to_utc(dt: datetime) -> datetime:
    if dt.tzinfo is None:
        return dt.replace(tzinfo=timezone.utc)
    return dt.astimezone(timezone.utc)


def build_event(
    notion_id: str,
    title: str,
    status_emoji: str,
    status_name: str,
    start_iso: Optional[str],
    end_iso: Optional[str],
    reminder_iso: Optional[str],
    description: Optional[str],
    *,
    category: Optional[str] = None,
    color: Optional[str] = None,
    url: Optional[str] = None,
) -> str:
    cal = Calendar()
    cal.add('prodid', '-//Notion Sync//EN')
    cal.add('version', '2.0')
    cal.add('calscale', 'GREGORIAN')

    event = Event()
    event.add('uid', build_uid(notion_id))
    summary_payload = (title or "").strip() or "Untitled"
    summary = f"{status_emoji}{summary_payload}"
    event.add('summary', summary)
    # Add per-event color when available (RFC 7986 COLOR)
    if color:
        event.add('color', color)
    if category:
        event.add('categories', [category])
    event.add('dtstamp', to_utc(datetime.now(timezone.utc)))
    event.add('last-modified', to_utc(datetime.now(timezone.utc)))

    if start_iso:
        if 'T' in start_iso:
            start = dtparser.isoparse(start_iso)
            event.add('dtstart', to_utc(start))
            if end_iso:
                end = dtparser.isoparse(end_iso)
            else:
                end = start + DEFAULT_TIMED_EVENT_DURATION
            event.add('dtend', to_utc(end))
        else:
            d = dtparser.isoparse(start_iso).date()
            event.add('dtstart', d)
            if end_iso:
                d2 = dtparser.isoparse(end_iso).date() + timedelta(days=1)
            else:
                d2 = d + timedelta(days=1)
            event.add('dtend', d2)

    event.add(
        'description',
        _compose_description(
            category=category,
            description=description,
        ),
    )
    event.add('url', url or f"https://www.notion.so/{notion_id.replace('-', '')}")

    if reminder_iso and start_iso and 'T' in start_iso:
        start = dtparser.isoparse(start_iso)
        reminder = dtparser.isoparse(reminder_iso)
        delta = to_utc(start) - to_utc(reminder)
        minutes_before = int(delta.total_seconds() // 60)
        if minutes_before > 0:
            alarm = Alarm()
            alarm.add('action', 'DISPLAY')
            alarm.add('trigger', timedelta(minutes=-minutes_before))
            alarm.add('description', f"Reminder: {title}")
            event.add_component(alarm)

    cal.add_component(event)
    return cal.to_ical().decode()


def parse_ics_minimal(ics_text: str) -> Dict[str, Optional[str]]:
    cal = Calendar.from_ical(ics_text)
    uid = None
    title = None
    status = None
    start_date = None
    end_date = None
    last_modified = None
    reminder_abs: Optional[str] = None
    is_placeholder: Optional[str] = None
    category: Optional[str] = None
    color: Optional[str] = None
    notion_description: Optional[str] = None
    for comp in cal.walk('VEVENT'):
        uid = str(comp.get('uid')) if comp.get('uid') else None
        summary = comp.get('summary')
        if summary is not None:
            s = str(summary)
            candidate_status, candidate_title = _extract_summary_status(s)
            title = candidate_title
            if candidate_status:
                status = candidate_status
        xph = comp.get('X-NOTION-PLACEHOLDER')
        if xph is not None:
            is_placeholder = '1'
        categories_prop = comp.get('categories')
        if categories_prop:
            if isinstance(categories_prop, list):
                if categories_prop:
                    category = str(categories_prop[0])
            else:
                category = str(categories_prop)
        color_prop = comp.get('color')
        if color_prop:
            color = str(color_prop)
        dtstart = comp.get('dtstart')
        dtend = comp.get('dtend')
        if dtstart:
            val = dtstart.dt
            if isinstance(val, datetime):
                start_date = to_utc(val).isoformat()
            else:
                start_date = val.isoformat()
        if dtend:
            val = dtend.dt
            if isinstance(val, datetime):
                end_date = to_utc(val).isoformat()
            else:
                end_date = val.isoformat()
        desc = comp.get('description')
        if desc:
            text = str(desc)
            headers, body = _parse_description_fields(text)
            category = headers.get('Category', category)
            if body is not None:
                notion_description = body
            elif 'Description' in headers and notion_description is None:
                notion_description = headers.get('Description')
            status = headers.get('Status', status)
        lm = comp.get('last-modified')
        if lm:
            val = lm.dt
            if isinstance(val, datetime):
                last_modified = to_utc(val).isoformat()
            else:
                try:
                    last_modified = val.isoformat()
                except Exception:
                    last_modified = None
        for sub in comp.subcomponents:
            try:
                if isinstance(sub, Alarm):
                    trig = sub.get('trigger')
                    if trig and isinstance(trig, str) and trig.startswith('-PT') and start_date:
                        start_dt = dtparser.isoparse(start_date)
                        import re
                        m = re.match(r"-PT(\d+)M", trig)
                        if m:
                            minutes = int(m.group(1))
                            reminder_abs = to_utc(start_dt) - timedelta(minutes=minutes)
                            reminder_abs = reminder_abs.isoformat()
            except Exception:
                pass
        break

    notion_id = None
    if uid and uid.startswith('notion-') and '@' in uid:
        notion_id = uid.split('@', 1)[0].replace('notion-', '')
    return {
        'notion_id': notion_id,
        'title': title,
        'status': status,
        'start_date': start_date,
        'end_date': end_date,
        'last_modified': last_modified,
        'reminder': reminder_abs,
        'is_placeholder': is_placeholder,
        'category': category,
        'description': notion_description,
        'color': color,
    }


def update_event_fields(existing_ics: str, *,
                        title: Optional[str] = None,
                        status_name: Optional[str] = None,
                        start_iso: Optional[str] = None,
                        end_iso: Optional[str] = None,
                        reminder_iso: Optional[str] = None,
                        notion_id: Optional[str] = None,
                        url: Optional[str] = None,
                        category: Optional[str] = None,
                        description: Optional[str] = None,
                        color: Optional[str] = None) -> str:
    cal = Calendar.from_ical(existing_ics)
    for comp in cal.walk('VEVENT'):
        if notion_id:
            comp['uid'] = build_uid(notion_id)
        if title is not None or status_name is not None:
            current_summary = str(comp.get('summary') or '')
            current_status, current_title = _extract_summary_status(current_summary)
            title_candidate = title if title is not None else current_title
            cleaned_title = _clean_summary_title(title_candidate)
            new_title = cleaned_title if cleaned_title else (title_candidate or "").strip()
            new_status = status_name if status_name is not None else current_status
            emoji = STATUS_EMOJI.get((new_status or "").strip(), "")
            if emoji and new_title:
                comp['summary'] = f"{emoji}{new_title}"
            elif emoji:
                comp['summary'] = emoji
            else:
                comp['summary'] = new_title
        # Update per-event color when provided
        if color is not None:
            if color:
                comp['color'] = color
            elif 'color' in comp:
                del comp['color']
        if start_iso is not None:
            if 'T' in start_iso:
                sdt = dtparser.isoparse(start_iso)
                comp['dtstart'] = to_utc(sdt)
                if end_iso is not None:
                    edt = dtparser.isoparse(end_iso) if end_iso else None
                    if edt:
                        comp['dtend'] = to_utc(edt)
            else:
                d = dtparser.isoparse(start_iso).date()
                comp['dtstart'] = d
                if end_iso is not None:
                    comp['dtend'] = dtparser.isoparse(end_iso).date() if end_iso else (d + timedelta(days=1))
        if category is not None:
            if category:
                comp['categories'] = [category]
            elif 'categories' in comp:
                del comp['categories']
        should_update_description = any(
            value is not None for value in (category, description)
        )
        if should_update_description:
            comp['description'] = _compose_description(
                category=category or '',
                description=description,
            )
        if url is not None:
            comp['url'] = url
        if reminder_iso is not None and start_iso and 'T' in start_iso:
            new_subs = []
            for sc in comp.subcomponents:
                if not isinstance(sc, Alarm):
                    new_subs.append(sc)
            comp.subcomponents = new_subs
            start = dtparser.isoparse(start_iso)
            reminder = dtparser.isoparse(reminder_iso)
            delta = to_utc(start) - to_utc(reminder)
            minutes_before = int(delta.total_seconds() // 60)
            if minutes_before > 0:
                alarm = Alarm()
                alarm.add('action', 'DISPLAY')
                alarm.add('trigger', f"-PT{minutes_before}M")
                alarm.add('description', f"Reminder: {title or comp.get('summary')}")
                comp.add_component(alarm)
        break
    return cal.to_ical().decode()


def normalize_from_notion(task) -> Dict[str, Optional[str]]:
    return {
        'title': getattr(task, 'title', None),
        'status': getattr(task, 'status', None),
        'category': getattr(task, 'category', None),
        'start_date': getattr(task, 'start_date', None),
        'end_date': getattr(task, 'end_date', None),
        'reminder': getattr(task, 'reminder', None),
        'description': (getattr(task, 'description', None) or None),
    }

File: src/calendar/ics.ts
import ICAL from "ical.js";
import ical from "ical-generator";
import { STATUS_CANONICAL_VARIANTS, STATUS_EMOJI_SETS } from "../sync/constants";

const DEFAULT_TIMED_EVENT_DURATION_MS = 60 * 60 * 1000; // 1 hour

// Map canonical Notion statuses to iCalendar VEVENT STATUS values (RFC 5545).
const NOTION_STATUS_TO_ICS: Record<string, string> = {
  Todo: "TENTATIVE",
  "In progress": "CONFIRMED",
  Completed: "CONFIRMED",
  Overdue: "CONFIRMED",
  Cancelled: "CANCELLED",
};

// Reverse: map iCalendar STATUS back to a canonical Notion status hint.
// This is a weak signal — the description Status header takes precedence.
const ICS_STATUS_TO_NOTION: Record<string, string> = {
  TENTATIVE: "Todo",
  CANCELLED: "Cancelled",
  // CONFIRMED is ambiguous (In progress or Completed), so we don't map it.
};

const emojiStatus = Object.fromEntries(
  Object.values(STATUS_EMOJI_SETS).flatMap((emojiSet) =>
    Object.entries(emojiSet).map(([canonical, emoji]) => [emoji.trim(), canonical]),
  ),
);

const statusPrefixesLower = [...new Set(
  Object.values(STATUS_CANONICAL_VARIANTS).flatMap((variants) => variants.map((variant) => variant.trim().toLowerCase())),
)].sort((left, right) => right.length - left.length);

export type ParsedIcs = {
  notionId: string | null;
  title: string | null;
  status: string | null;
  startDate: string | null;
  endDate: string | null;
  lastModified: string | null;
  reminder: string | null;
  isPlaceholder: string | null;
  category: string | null;
  description: string | null;
  color: string | null;
  url: string | null;
};

export function buildUid(notionId: string): string {
  return `notion-${notionId}@sync`;
}

export function buildEvent(input: {
  notionId: string;
  title: string;
  statusEmoji: string;
  statusName: string;
  startIso: string | null;
  endIso: string | null;
  reminderIso: string | null;
  description: string | null;
  category?: string | null;
  color?: string | null;
  url?: string | null;
  lastModified?: string | null;
}): string {
  const calendar = ical({
    prodId: { company: "Notion Sync", product: "Notion Sync" },
    method: "PUBLISH" as any,
    name: "Notion",
  });

  const event = calendar.createEvent({
    start: new Date(),
    id: buildUid(input.notionId),
    summary: input.statusEmoji?.trim()
      ? `${input.statusEmoji.trim()} ${(input.title || "").trim() || "Untitled"}`
      : (input.title || "").trim() || "Untitled",
    description: composeDescription({
      category: input.category || null,
      description: input.description || null,
    }),
    url: input.url || `https://www.notion.so/${input.notionId.replaceAll("-", "")}`,
  });
  (event as any).stamp(new Date());
  const lastModDate = input.lastModified ? new Date(input.lastModified) : new Date();
  (event as any).lastModified(Number.isNaN(lastModDate.getTime()) ? new Date() : lastModDate);
  if (input.category) {
    (event as any).categories([{ name: input.category }]);
  }

  // Set iCalendar STATUS property based on Notion task status.
  const icsStatus = NOTION_STATUS_TO_ICS[input.statusName];
  if (icsStatus) {
    (event as any).status(icsStatus);
  }

  if (input.startIso) {
    if (input.startIso.includes("T")) {
      const start = new Date(input.startIso);
      event.start(start);
      event.end(input.endIso ? new Date(input.endIso) : new Date(start.getTime() + DEFAULT_TIMED_EVENT_DURATION_MS));
    } else {
      const start = parseDateOnly(input.startIso);
      if (start) {
        event.allDay(true);
        event.start(start);
        const endDate = input.endIso ? parseDateOnly(input.endIso) : null;
        const inclusiveEnd = endDate || start;
        event.end(new Date(inclusiveEnd.getTime() + 24 * 60 * 60 * 1000));
      }
    }
  }

  if (input.reminderIso && input.startIso?.includes("T")) {
    const start = new Date(input.startIso);
    const reminder = new Date(input.reminderIso);
    const minutesBefore = Math.floor((start.getTime() - reminder.getTime()) / (60 * 1000));
    if (minutesBefore > 0) {
      event.createAlarm({
        type: "display",
        trigger: minutesBefore * 60,
        description: `Reminder: ${input.title}`,
      } as any);
    }
  }
  let output = calendar.toString();
  // Inject X-NOTION-PAGE-ID custom property for reliable round-trip identification.
  // This survives server-side UID normalization and filename changes.
  output = output.replace(/(UID:[^\r\n]+\r\n)/, `$1X-NOTION-PAGE-ID:${input.notionId}\r\n`);
  if (input.color) {
    output = output.replace(/(SUMMARY:[^\r\n]+\r\n)/, `$1COLOR:${input.color}\r\n`);
  }
  return output;
}

export function parseIcsMinimal(icsText: string): ParsedIcs {
  const jcalData = ICAL.parse(icsText);
  const component = new ICAL.Component(jcalData);
  const vevent = component.getFirstSubcomponent("vevent");
  if (!vevent) {
    return emptyParsed();
  }
  const event = new ICAL.Event(vevent);

  const { status: summaryStatus, title } = extractSummaryStatus(event.summary || "");
  let status = summaryStatus;
  let category = firstText(arrayify(event.component.getFirstPropertyValue("categories")));
  const color = normalizeText(event.component.getFirstPropertyValue("color"));
  const rawDescription = normalizeText(event.description);
  // Normalize whitespace: CalDAV servers may add \r\n; canonicalize to \n.
  const descriptionValue = rawDescription ? rawDescription.replace(/\r\n/g, "\n").replace(/\r/g, "\n") : null;
  let description: string | null = null;
  let url = normalizeText(event.component.getFirstPropertyValue("url"));

  if (descriptionValue) {
    const parsed = parseDescriptionFields(descriptionValue);
    category = parsed.headers.Category || category;
    // Only use the user's actual description body, not the structured metadata
    // headers (Source, Status, Notion URL, etc.) that we embed in the ICS.
    // If there's no body and no explicit Description header, the description
    // was purely metadata — return null to match Notion's empty description.
    description = parsed.body || parsed.headers.Description || null;
    if (!summaryStatus || summaryStatus === "Overdue") {
      status = parsed.headers.Status || status;
    }
  }

  // Fall back to iCalendar STATUS property if no status was extracted from
  // the summary emoji or the description headers.
  if (!status) {
    const veventStatus = normalizeText(vevent.getFirstPropertyValue("status"));
    if (veventStatus) {
      status = ICS_STATUS_TO_NOTION[veventStatus.toUpperCase()] || null;
    }
  }

  const lastModifiedValue = event.component.getFirstPropertyValue("last-modified") as ICAL.Time | null;
  const lastModified = lastModifiedValue ? toIso(lastModifiedValue) : null;
  const startValue = event.startDate;
  const endValue = event.endDate;
  const startDate = startValue ? toIso(startValue, false) : null;
  let endDate = endValue ? toIso(endValue, event.startDate?.isDate || false, true) : null;
  // For all-day events: if the parsed end date equals the start date, the
  // original event was a single-day event with no explicit end. Return null
  // to match Notion's representation and avoid hash mismatches.
  if (endDate && startDate && event.startDate?.isDate && endDate === startDate) {
    endDate = null;
  }

  const alarms = vevent.getAllSubcomponents("valarm") || [];
  let reminder: string | null = null;
  for (const alarm of alarms) {
    const trigger = alarm.getFirstPropertyValue("trigger");
    if (!trigger || !startDate) {
      continue;
    }
    const start = new Date(startDate);
    if (trigger instanceof ICAL.Duration) {
      const seconds = trigger.toSeconds();
      reminder = new Date(start.getTime() + seconds * 1000).toISOString();
      break;
    }
  }

  const uid = normalizeText(event.uid);
  // Prefer X-NOTION-PAGE-ID custom property for reliable identification.
  // Falls back to extracting from UID (notion-{pageId}@sync).
  const xNotionPageId = normalizeText(vevent.getFirstPropertyValue("x-notion-page-id"));
  const notionId = xNotionPageId
    || (uid && uid.startsWith("notion-") && uid.includes("@")
        ? uid.split("@", 1)[0].replace("notion-", "")
        : null);
  const placeholder = vevent.getFirstPropertyValue("x-notion-placeholder") ? "1" : null;

  return {
    notionId,
    title,
    status,
    startDate,
    endDate,
    lastModified,
    reminder,
    isPlaceholder: placeholder,
    category,
    description,
    color,
    url,
  };
}

export function normalizeFromNotion(task: {
  title?: string | null;
  status?: string | null;
  category?: string | null;
  startDate?: string | null;
  endDate?: string | null;
  reminder?: string | null;
  description?: string | null;
}): Record<string, string | null> {
  return {
    title: task.title || null,
    status: task.status || null,
    category: task.category || null,
    startDate: task.startDate || null,
    endDate: task.endDate || null,
    reminder: task.reminder || null,
    description: task.description || null,
  };
}

function composeDescription(input: { category: string | null; description: string | null }): string {
  if (input.description) {
    return input.description;
  }
  if (input.category) {
    return `Category: ${input.category}`;
  }
  return "";
}

function extractSummaryStatus(summary: string): { status: string | null; title: string } {
  if (!summary) {
    return { status: null, title: "" };
  }
  const [head, ...tailParts] = summary.split(" ");
  const headStatus = emojiStatus[head?.trim() || ""];
  if (headStatus) {
    return { status: headStatus, title: tailParts.join(" ").trim() };
  }
  const stripped = summary.trim();
  const summaryStatus = emojiStatus[stripped];
  if (summaryStatus) {
    return { status: summaryStatus, title: "" };
  }
  const firstCharStatus = emojiStatus[summary[0] || ""];
  if (firstCharStatus) {
    return { status: firstCharStatus, title: summary.slice(1).trim() };
  }
  return { status: null, title: summary };
}

// Known metadata header keys that we embed in the ICS description.
// Only these are extracted as structured headers; everything else is
// treated as user-authored body text to preserve round-trip fidelity.
const KNOWN_HEADER_KEYS = new Set(["Source", "Status", "Notion URL", "Category", "Description"]);

function parseDescriptionFields(text: string): {
  headers: Record<string, string>;
  body: string | null;
} {
  const headers: Record<string, string> = {};
  let body: string | null = null;
  let headerText = text;
  if (text.includes("\n\n")) {
    const split = text.split("\n\n", 2);
    headerText = split[0] || "";
    body = split[1]?.trim() || null;
  }
  const headerCandidates = headerText.includes("\n")
    ? headerText.split("\n").map((line) => line.trim()).filter(Boolean)
    : headerText.split("|").map((line) => line.trim()).filter(Boolean);

  // Collect lines that don't match a known header — these are user content
  // that ended up before the paragraph break.
  const unknownLines: string[] = [];
  for (const item of headerCandidates) {
    const separator = item.indexOf(":");
    if (separator <= 0) {
      unknownLines.push(item);
      continue;
    }
    const key = item.slice(0, separator).trim();
    const value = item.slice(separator + 1).trim();
    if (KNOWN_HEADER_KEYS.has(key)) {
      headers[key] = value;
    } else {
      // Not a known metadata header — treat the whole line as body content.
      unknownLines.push(item);
    }
  }

  // Prepend any unknown lines to the body so user content isn't lost.
  if (unknownLines.length > 0) {
    const unknownText = unknownLines.join("\n").trim();
    body = body ? `${unknownText}\n\n${body}` : unknownText;
  }

  if (!body && headers.Description) {
    body = headers.Description;
  }
  // If no structured headers were found and no body was extracted,
  // the entire text is plain description content — not metadata.
  if (!body && Object.keys(headers).length === 0) {
    body = text.trim() || null;
  }
  return { headers, body };
}

function parseDateOnly(value: string): Date | null {
  const [yearText, monthText, dayText] = value.split("-");
  const year = Number.parseInt(yearText || "", 10);
  const month = Number.parseInt(monthText || "", 10);
  const day = Number.parseInt(dayText || "", 10);
  if (!Number.isFinite(year) || !Number.isFinite(month) || !Number.isFinite(day)) {
    return null;
  }
  return new Date(Date.UTC(year, month - 1, day));
}

function toIso(time: ICAL.Time, isDateRange = false, inclusiveDateEnd = false): string | null {
  if (time.isDate) {
    let year = time.year;
    let month = time.month;
    let day = time.day;
    if (inclusiveDateEnd) {
      const adjusted = new Date(Date.UTC(year, month - 1, day));
      adjusted.setUTCDate(adjusted.getUTCDate() - 1);
      year = adjusted.getUTCFullYear();
      month = adjusted.getUTCMonth() + 1;
      day = adjusted.getUTCDate();
    }
    return `${String(year).padStart(4, "0")}-${String(month).padStart(2, "0")}-${String(day).padStart(2, "0")}`;
  }
  return time.toJSDate().toISOString();
}

function arrayify<T>(value: T | T[] | null | undefined): T[] {
  if (value == null) {
    return [];
  }
  return Array.isArray(value) ? value : [value];
}

function firstText(values: unknown[]): string | null {
  for (const value of values) {
    const text = normalizeText(value);
    if (text) {
      return text;
    }
  }
  return null;
}

function normalizeText(value: unknown): string | null {
  if (typeof value !== "string") {
    return null;
  }
  const normalized = value.trim();
  return normalized || null;
}

function emptyParsed(): ParsedIcs {
  return {
    notionId: null,
    title: null,
    status: null,
    startDate: null,
    endDate: null,
    lastModified: null,
    reminder: null,
    isPlaceholder: null,
    category: null,
    description: null,
    color: null,
    url: null,
  };
}
